beacon-chain/blockchain/service.go

```diff
@@ -218,17 +179,126 @@ func (c *ChainService) doesPoWBlockExist(block *types.Block) bool {
 	return powBlock != nil
 }
 
+func (c *ChainService) updateHead(slotInterval <-chan time.Time, done <-chan struct{}) {
+	for {
+		select {
+		case <-done:
+			return
+		case <-slotInterval:
+			c.currentSlot++
+
+			log.WithField("slotNumber", c.currentSlot).Info("New beacon slot")
+
+			// First, we check if there were any blocks processed in the previous slot.
+			// If there is, we fetch the first one from the DB.
+			if len(c.blocksPendingProcessing) == 0 {
+				continue
+			}
+
+			// Naive fork choice rule: we pick the first block we processed for the previous slot
+			// as canonical.
+			// TODO: Use a better lookup key or abstract into a GetBlock method from DB.
+			hasBlock, err := c.beaconDB.Has(c.blocksPendingProcessing[0])
+			if err != nil {
+				continue
+			}
+			if !hasBlock {
+				continue
+			}
+			encoded, err := c.beaconDB.Get(c.blocksPendingProcessing[0])
+			if err != nil {
+				continue
+			}
+			blockProto := &pb.BeaconBlock{}
+			if err := proto.Unmarshal(encoded, blockProto); err != nil {
+				continue
+			}
+			block := types.NewBlock(blockProto)
+			h, err := block.Hash()
+			if err != nil {
+				log.Errorf("Could not hash incoming block: %v", err)
+			}
+
+			log.Info("Applying fork choice rule")
+			aState := c.chain.ActiveState()
+			cState := c.chain.CrystallizedState()
+			isTransition := cState.IsCycleTransition(c.currentSlot - 1)
+
+			if isTransition {
+				cState, err = cState.CalculateNewCrystallizedState(aState, block.SlotNumber())
+				if err != nil {
+					log.Errorf("Initialize new cycle transition failed: %v", err)
+				}
+			}
+
+			parentBlock, err := c.chain.getBlock(block.ParentHash())
+			if err != nil {
+				log.Errorf("Failed to get parent slot of block 0x%x", h)
+				continue
+			}
+			aState, err = aState.CalculateNewActiveState(block, cState, parentBlock.SlotNumber())
+			if err != nil {
+				log.Errorf("Compute active state failed: %v", err)
+				continue
+			}
+
+			if err := c.chain.SetActiveState(aState); err != nil {
+				log.Errorf("Write active state to disk failed: %v", err)
+			}
+
+			if err := c.chain.SetCrystallizedState(cState); err != nil {
+				log.Errorf("Write crystallized state to disk failed: %v", err)
+			}
+
+			vals, err := casper.ShuffleValidatorsToCommittees(
```

Comment:
We don't need this here since shuffling happens inside `CalculateNewCrystallizedState`